### PR TITLE
Separated Retriever as Base abstract class and custom post function for API

### PR DESCRIPTION
**Pull request by S1LV3RJ1NX**
- Created a base class `TFBaseRetriever` that will be inherited whenever a user needs to make a new retriever
- Added a sample `CreditCardRetriver` class that inherits `TFBaseRetriever` and `BaseRetriever` of langchain for document retrieval of credit cards
- Created a custom decorator `@retriever_post` function of FastAPI, that overrides `post` function of FastAPI, to generate API routes in the format: `/retriver/<RETRIEVER_NAME>/<func_name>`

### CODE REVIEW

File: backend/modules/retrievers/credit_card_retriver.bak.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+from typing import List
```
Comment (innoavator) [inline backend/modules/retrievers/credit_card_retriver.bak.py:1]: Lets remove the bak file

File: backend/modules/retrievers/base.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    collection_name: str = Field(
+        default=None,
+        title="Collection name on which to search",
+    )
+
+    query: str = Field(title="Query using which the similar documents will be searched", max_length=1000)
+
+
+class TFQueryEngine(ABC):
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:39]: Add more documentation

File: backend/modules/retrievers/base.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    @abstractmethod
+    def get_embeddings(self, collection_name: str):
+        """Logic to get embeddings"""
+    
+    @abstractmethod
+    def get_vector_store(self, collection_name: str):
+        """Logic to get vector store"""
+    
+    @abstractmethod
+    def get_llm(self):
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:66]: Make it fully typed

**Review by innoavator — CHANGES_REQUESTED** [review]
Lets fix this PR properly to make it very clear - the PR is not easy to understand as of now.

File: backend/server/app.py [hunk 4]
```diff
@@ -243,76 +243,76 @@ async def get_collection_status(collection_name: str = Path(title="Collection na
...
+
+#         return {
+#             "answer": outputs["result"],
+#             "docs": outputs.get("source_documents") or [],
+#         }
+#     except HTTPException as exp:
+#         raise exp
+#     except Exception as exp:
+#         logger.exception(exp)
+#         raise HTTPException(status_code=500, detail=str(exp))
```
[the remark this answers is not in the document]
Comment (S1LV3RJ1NX) [reply backend/server/app.py:315]: Addressed.

File: backend/modules/retrievers/base.py (hunk 3 above)
Comment (S1LV3RJ1NX) [reply backend/modules/retrievers/base.py:66]: Resolved.

File: backend/modules/retrievers/base.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        3. Get the necessary embeddings
+        4. Define an LLM to format the query and retrival
+        5. Define retriver logic
+        6. Define query, this is a compulsory method, with compulsory collection name argument, either you can write your custom 
+            logic here or use the above helper functions to stitch together your query engine components.
+    """
+
+    retriever_name: str = ''
+
+    @classmethod
```
Comment (akashg3627) [inline backend/modules/retrievers/base.py:54]: abstractmethod
-> _get_collection, _get_vector_store_client, _get_embeddings, _get_llm 

query, get_docs

File: backend/modules/retrievers/base.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        4. Define an LLM to format the query and retrival
+        5. Define retriver logic
+        6. Define query, this is a compulsory method, with compulsory collection name argument, either you can write your custom 
+            logic here or use the above helper functions to stitch together your query engine components.
+    """
+
+    retriever_name: str = ''
+
+    @staticmethod
+    def _get_collection(collection_name: str) -> Collection:
```
Comment (akashg3627) [inline backend/modules/retrievers/base.py:89]:
```suggestion
    @abstractmethod
    def _get_collection(self,collection_name: str) -> Collection:
```

File: backend/modules/retrievers/base.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        1. Get your collection by giving the collection name
+        2. Get the corresponding vector store client and vector store
+        3. Get the necessary embeddings
+        4. Define an LLM to format the query and retrival
+        5. Define retriver logic
+        6. Define query, this is a compulsory method, with compulsory collection name argument, either you can write your custom 
+            logic here or use the above helper functions to stitch together your query engine components.
+    """
+
+    def _get_vector_store(self, collection_name: str) -> VectorStore:
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:86]: get_vector_store_for_collection

File: backend/modules/retrievers/base.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    by indicating any uncertainty or lack of knowledge regarding the correct answer to avoid providing incorrect information.""",
+        title="System prompt to use for generating answer to the question",
+    )
+    prompt_template: Optional[str] = Field(
+        default="""Here is the context information:\n\n'''\n{context}\n'''\n\nQuestion: {question}\nAnswer:""",
+        title="Prompt Template to use for generating answer to the question using the context",
+    )
+
+
+class BaseQueryEngine:
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:72]: LangchainQueryEngine

File: backend/modules/retrievers/base.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
+from abc import ABC, abstractmethod
+from pydantic import BaseModel, Field
+from typing import List, Literal, Optional, Any
+from fastapi import APIRouter
+
+from backend.settings import settings
+from backend.utils.logger import logger
+from backend.modules.metadata_store.models import Collection
+from backend.utils.base import RetrieverConfig, LLMConfig
+from servicefoundry.langchain import TrueFoundryChat
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:10]: Let's remove this import

File: backend/modules/retrievers/base.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        # Get the embeddings from the collections embedding config
+        embedding = get_embedder(collection.embedder_config)
+
+        # Get vector store for the corresponding embedding
+        vector_store = vector_store_client.get_vector_store(embedding)
+
+        return vector_store
+    
+
+    def _get_llm(self, model_config: LLMConfig, system_prompt: str) -> TrueFoundryChat:
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:107]: import openai client - return langchain.llm

File: backend/modules/retrievers/base.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        """Logic for using a custom llm if any"""
+        llm = TrueFoundryChat(
+            model=model_config.name,
+            model_parameters=model_config.parameters,
+            system_prompt=system_prompt,
+        )
+        return llm
+
+    def _get_retriever(self, collection_name: str, retriever_config: RetrieverConfig) -> VectorStoreRetriever:
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:116]: langchain.retriever.VectorStoreRetriever

File: backend/modules/retrievers/base.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class RAGEngine(ABC):
+
+    retriever_name: str = ''    
+    query_object: BaseQueryEngine = None
+
+    @staticmethod
+    @abstractmethod
+    def get_documents(input: QueryInput) -> List[Document]:
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:137]: Let's remove this

File: backend/modules/retrievers/credit_card_retriver.py [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    RAGEngine, 
+    QueryInput, 
+    retriever_post
+)
+
+
+
+RETRIEVER_NAME = 'credit-card'
+
+class CredCardDocsQuery(QueryInput):
```
Comment (innoavator) [inline backend/modules/retrievers/credit_card_retriver.py:21]: Make this DefaultRetriever

File: backend/utils/base.py [hunk 14]
```diff
@@ -207,6 +207,32 @@ class SearchQuery(BaseModel):
     )
 
 
+class RetrivalQuery(BaseModel):
```
Comment (innoavator) [inline backend/utils/base.py:210]: Remove

**Review by innoavator — CHANGES_REQUESTED** [review]

File: backend/modules/retrievers/default_retriever.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+                input_variables=["context", "question"],
+                template=input.prompt_template,
+            )
+            
+            # Define reteiver and llm
+            retriever = query_object._get_retriever(input.collection_name, input.retriever_config)
+            llm = query_object._get_llm(input.model_configuration, input.system_prompt)
+
+            # Define retrieval chain
+            retrieval_chain = get_retrieval_chain(
```
Comment (akashg3627) [inline backend/modules/retrievers/default_retriever.py:85]: lets remove this function and directly use RetrivalQA class here

File: backend/modules/retrievers/base.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    )
+
+    retriever_config: Optional[RetrieverConfig] = Field(
+        title="Retriever configuration",
+    )
+
+    model_configuration: Optional[LLMConfig]
+
+    system_prompt: Optional[str] = Field(
+        default="""Your task is to craft the most helpful, highly informative, accurate and comprehensive answers possible,
```
Comment (akashg3627) [inline backend/modules/retrievers/base.py:57]: lets move these texts to some common constants file in src folder

File: backend/modules/retrievers/base.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    ensuring they are easy to understand and implement. Use the context information provided as a reference to form accurate responses 
+    that incorporate as much relevant detail as possible. Strive to make each answer clear and precise to enhance user comprehension and 
+    assist in solving their problems effectively.\n\nEmploy the provided context information meticulously to craft precise answers, 
+    ensuring they incorporate all pertinent details. Structure your responses for ease of reading and relevance by providing as much as 
+    information regarding it. Make sure the answers are well detailed and provide proper references. Align your answers with the context given and maintain transparency 
+    by indicating any uncertainty or lack of knowledge regarding the correct answer to avoid providing incorrect information.""",
+        title="System prompt to use for generating answer to the question",
+    )
+    prompt_template: Optional[str] = Field(
+        default="""Here is the context information:\n\n'''\n{context}\n'''\n\nQuestion: {question}\nAnswer:""",
```
Comment (akashg3627) [inline backend/modules/retrievers/base.py:67]: lets move these texts to some common constants file in src folder

File: backend/modules/retrievers/base.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        embedding = get_embedder(collection.embedder_config)
+
+        # Get vector store for the corresponding embedding
+        vector_store = vector_store_client.get_vector_store(embedding)
+
+        return vector_store
+    
+
+    def _get_llm(self, model_config: LLMConfig, system_prompt: str) -> LangChainSimpleChatModel:
+        """Logic for using a custom llm if any"""
```
Comment (akashg3627) [inline backend/modules/retrievers/base.py:108]:
```suggestion
        """Function to get langchain chat llm class"""
```

File: backend/modules/retrievers/base.py [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+class QueryInput(BaseModel):
+
+    collection_name: str = Field(
+        default=None,
+        title="Collection name on which to search",
+    )
+
+    query: str = Field(title="Query using which the similar documents will be searched", max_length=1000)
+
+    retrieval_chain_name: Optional[Literal["RetrievalQA"]] = Field(
```
Comment (akashg3627) [inline backend/modules/retrievers/base.py:41]: not needed

File: backend/modules/retrievers/base.py [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+        default="RetrievalQA",
+        title="Name of the retrieval chain to use for retrieving documents",
+    )
+
+    retriever_config: Optional[RetrieverConfig] = Field(
+        title="Retriever configuration",
+    )
+
+    model_configuration: Optional[LLMConfig | str]
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:50]: Let's remove the str?

File: backend/modules/retrievers/base.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+
+    prompt_template: Optional[str] = Field(
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:52]: Can we have a better default prompt?

File: backend/modules/retrievers/base.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+        default="""Here is the context information:\n\n'''\n{context}\n'''\n\nQuestion: {question}\nAnswer:""",
+        title="Prompt Template to use for generating answer to the question using the context",
+    )
+
+
+class LangchainQueryEngine:
+    """Base Query Engine class, this class lets you write your own retriver. All it's methods can be inherited and overridden
```
Comment (innoavator) [inline backend/modules/retrievers/base.py:59]: nit: retriever

**Review by innoavator — CHANGES_REQUESTED** [review]

File: backend/modules/retrievers/default_retriever.py [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+
+class DefaultRAGTool(BaseRAGTool):
+
+    retriever_name: str = RETRIEVER_NAME
+    
+    @retriever_post(RETRIEVER_NAME)
+    def query(input: DefaultInputQuery):    
+        """We implement query function for getting answer and relavant documents"""
+        try:
+            from langchain.prompts import PromptTemplate
```
Comment (innoavator) [inline backend/modules/retrievers/default_retriever.py:31]: Can we not import within the function?

**Review by innoavator — CHANGES_REQUESTED** [review]

File: backend/modules/retrievers/__init__.py [hunk 24]
```diff
@@ -18,12 +18,12 @@ def get_retriever(
     )
     SUPPORTED_RETRIEVERS = [
         "VectorStoreRetriever",
-        "CustomRetriever",
+        "DefaultRAGTool",
```
Comment (innoavator) [inline backend/modules/retrievers/__init__.py:21]: This should be a registry instead of adding it in a central place

**Review by innoavator — CHANGES_REQUESTED** [review]

File: backend/modules/retrievers/__init__.py (hunk 24 above)
Comment (akashg3627) [inline backend/modules/retrievers/__init__.py:21]: Lets get this part of code deleted

Comment (S1LV3RJ1NX) [reply backend/modules/retrievers/__init__.py:21]: To be done later...deferred for now.

**Pull request closed without merging**